Scope reusing & smaller scopes

yosiat opened the pull request:
**This pull request, is experiment, If you like the idea, we can improve the readability and the quality of the code**

For each expression we are creating "scope pool", which is object pool of scopes - with some extra magic.
By doing quick analysis on the node AST I know which tags and fields he requires. so we put only the required ones.
For example: "value" > 10, I fill only "value" from field or tag.
```
name                     old time/op    new time/op    delta
_T10_P500_AlertTask-4       133ms ± 4%     123ms ± 4%   -7.58%  (p=0.008 n=5+5)
_T10_P50000_AlertTask-4     13.4s ± 8%     12.3s ± 7%     ~     (p=0.056 n=5+5)
_T1000_P500_AlertTask-4     13.5s ± 4%     12.1s ± 3%  -10.46%  (p=0.008 n=5+5)

name                     old alloc/op   new alloc/op   delta
_T10_P500_AlertTask-4      32.2MB ± 0%    26.0MB ± 0%  -19.32%  (p=0.008 n=5+5)
_T10_P50000_AlertTask-4    3.26GB ± 0%    2.62GB ± 0%  -19.71%  (p=0.008 n=5+5)
_T1000_P500_AlertTask-4    3.21GB ± 0%    2.61GB ± 0%  -18.56%  (p=0.008 n=5+5)

name                     old allocs/op  new allocs/op  delta
_T10_P500_AlertTask-4        408k ± 0%      335k ± 0%  -17.85%  (p=0.008 n=5+5)
_T10_P50000_AlertTask-4     41.5M ± 0%     34.1M ± 0%  -17.98%  (p=0.008 n=5+5)
_T1000_P500_AlertTask-4     40.2M ± 0%     33.1M ± 0%  -17.61%  (p=0.008 n=5+5)
```

I thought about this idea while researching the performance of alerts, but before that I wanted to implement "compiled stateful expression" ( #491 ).
If we combine this pull request with #491 , we will have great performance and low memory usage while evaluating predicates.

nathanielc commented:
@yosiat Thanks! I was hoping this would also get done. At a quick glance it looks good to go once #491 gets merged.

yosiat commented:
@nathanielc I think we need to define how it will be used.
I currently stuck in StatefulExpression but it might be more clean and ok to be it in it's own struct that StatefulExpression will own?

nathanielc commented:
>I currently stuck in StatefulExpression but it might be more clean and ok to be it in it's own struct that StatefulExpression will own?

I think having it on its own would be good.

yosiat commented:
@nathanielc I will fix that tomorrow and this PR will be ready to merge; I am going to sleep 💤

nathanielc commented:
@yosiat yes, please get some rest :sleeping:

yosiat commented:
@nathanielc I finished this PR 😄

- Refactored the code and the tests
- Rebased with the new expressions

**Benchmarks (compared to master with new stateful expressions)**
_I updated the benchmarks in the pull request as well_
Ran with count=5
```
name                     old time/op    new time/op    delta
_T10_P500_AlertTask-4       133ms ± 4%     123ms ± 4%   -7.58%  (p=0.008 n=5+5)
_T10_P50000_AlertTask-4     13.4s ± 8%     12.3s ± 7%     ~     (p=0.056 n=5+5)
_T1000_P500_AlertTask-4     13.5s ± 4%     12.1s ± 3%  -10.46%  (p=0.008 n=5+5)

name                     old alloc/op   new alloc/op   delta
_T10_P500_AlertTask-4      32.2MB ± 0%    26.0MB ± 0%  -19.32%  (p=0.008 n=5+5)
_T10_P50000_AlertTask-4    3.26GB ± 0%    2.62GB ± 0%  -19.71%  (p=0.008 n=5+5)
_T1000_P500_AlertTask-4    3.21GB ± 0%    2.61GB ± 0%  -18.56%  (p=0.008 n=5+5)

name                     old allocs/op  new allocs/op  delta
_T10_P500_AlertTask-4        408k ± 0%      335k ± 0%  -17.85%  (p=0.008 n=5+5)
_T10_P50000_AlertTask-4     41.5M ± 0%     34.1M ± 0%  -17.98%  (p=0.008 n=5+5)
_T1000_P500_AlertTask-4     40.2M ± 0%     33.1M ± 0%  -17.61%  (p=0.008 n=5+5)
```

I think we can do the same for eval.go using some pooling their, what do you think? do we have good test coverage for what's happening there?

nathanielc commented:
@yosiat Looks good. Can we also update the `eval.go` usage to use the scope pool? Basically any place that use `mergeFieldsAndTags` should be updated and then I think we can remove the `mergeFieldsAndTags` function.

yosiat commented:
@nathanielc It's a bit problematic there because we have multiple expressions, we can create at startup "combined reference variables" set and use it, what do you think?

nathanielc commented:
@yosiat I don't believe the scope is shared across the expressions, so they should all be able to operate independently.

yosiat commented:
@nathanielc Looking at the code, looks like they use the same "vars", and after evaluation you are updating the scope..

It looks like the second expression is using the result of the first expression, etc.

nathanielc commented:
@yosiat Yes, the `scope` cascades across expressions but the `state` is independent, and the set of referenced values is constant as well.

You should be able to use `ScopePool.Get` and check if the next expression even wants the result of the previous expression.  In other words create a scope for each expression and then only cascade needed values. The currently implementation is lazy and just reuses the same scope.

yosiat commented:
umpf, so why not simply on upstart get the reference variables used in all scopes and merge them to one set? so the initial scope will be the merged? in term of performance your approach will win, but this approach will be much simpler.

nathanielc commented:
Go with the simpler approach :) The biggest win is to remove allocations for each evaluation of the expression. Other than that lets keep it simple.

yosiat commented:
What do you think about creating "MergedScopePool" which will accept list of nodes and and create scope pool for them?

nathanielc commented:
I would rather not create another type. 

What if the `NewScope` method took a list of names `[]string` instead of a node. Then you could easily create ScopePools for difference use cases. You would need to expose a way to get the referencedVariables still. I think adding a method `ReferencedVariables` to the `StatefulExpression` would work as it has a reference to the node already. Then remove the `ScopePool` method from the `StatefulExpression`, in other words let the caller use the pool but don't require it. Then in `eval.go` on upstart you would call `ReferencedVariable` for each expression and merge them and create a single ScopePool for all the expressions. Does that makes sense? Basically instead of pushing the merge logic down into the `stateful` package leave it in the `eval.go` code since its the code that needs it.

yosiat commented:
@nathanielc liked your idea, I will start to play with it and update ~

yosiat commented:
@nathanielc I will make "findReferenceVariables" a function on top of Node and will put under "tick/reflect.go"

yosiat commented:
@nathanielc after thinking a bit, if the StatefulExpression won't own the scope pool, it means that:

- alert: will need to contain a symmetric array for holding the scope pools
- EvalPredicate: will need to accept the scope pool

Are we ok with those changes?

nathanielc commented:
@yosiat +1

yosiat commented:
@nathanielc This is getting some what tedious, In where, alert and eval we have to "maintain" symmetric array for holding the scope pools. I think it might hear readability.

nathanielc commented:
@yosiat Then maybe you can supply the ScopePool to the StatefulExpression and it does the `maintaining`? But it still allows for the use case of the pool to be different.

yosiat commented:
@nathanielc I thought about it, but then it will be too much weird, because they will some dependency.

In order to create ScopePool you will need reference variables, which you get by calling "ReferenceVariables" on StatefulExpression.

nathanielc commented:
@yosiat OK, you are more in the details at this point. Do what makes sense...

yosiat commented:
@nathanielc What do you think? I haven't integrated it in eval.go yet..

nathanielc commented:
@yosiat 

>In order to create ScopePool you will need reference variables, which you get by calling "ReferenceVariables" on StatefulExpression.

This is essentially the only difference between alert/where and eval, is how the scope is created. It is used in the same manner. So either have a little duplicate code in alert and where or add helper functions to the `stateful` package. Something like `CreateScopePoolForExpressions` and it takes a list of expressions. Then eval passes in a list and where/alert pass in a list of one.

yosiat commented:
@nathanielc do we want "CreateScopePoolForExpressions" to replace "NewScope" or to be helper method in stateful/scope_pool.go? (I think the latter)

nathanielc commented:
@yosiat  The later, if something wants to create a scope in yet another different way it is still possible if we leave NewScope as is. Also that is why I called it `Create...` because there is implied logic happening not simply instantiating a new object.

yosiat commented:
@nathanielc I started implementing the `CreateScopePoolForExpressions` in eval.go, and it looks like we have another problem.
We need `ReferenceVariables` for `fillScopePool (in expr.go)` and in eval.go we will need to have another list for it.

I think instead of `CreateScopePoolForExpressions` we will create: `FindReferenceVariablesForNodes` and remove it from Expression.

So the usage will be something like:
```go
var node tick.Node // let's say we have node
refVariables := stateful.FindReferenceVariablesForNodes(node)
stateful.NewScopePool(refVariables)
```

and in eval.go, it will something like this:
```go
refVariables := stateful.FindReferenceVariablesForNodes(n.Expressions...)
stateful.NewScopePool(refVariables)
```

yosiat commented:
@nathanielc It's finished, squashed & updated benchmarks.

nathanielc commented on expr.go:
```diff
@@ -10,29 +10,44 @@ import (
+func fillScope(vars *tick.Scope, referenceVariables []string, now time.Time, fields models.Fields, tags models.Tags) error {
+	for _, refVariableName := range referenceVariables {
+		if refVariableName == "time" {
+			vars.Set("time", now.Local())
```
Should we `break` after this line? That way we don't accidentally set a field value named `time`.

yosiat replied:
Agree, but not `break`, `continue`

nathanielc replied:
Ah yes :)

nathanielc commented:
@yosiat :+1:  I'll merge on green. Since nothing user facing changes no need to write a CHANGELOG.
